### ISSUE TITLE
Do not render the whole table when interacting with a single row

### ISSUE THREAD

**Issue by StallionV** [ant-design/ant-design#23763]
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/selection-and-operation-ant-design-demo-r8d25?file=/index.js)

### Steps to reproduce
Add a Table
Use the `onRow` function or turn on rowSelections using `rowSelection`

### What is expected?
Only the rows being interacted with should be re-rendered.

### What is actually happening?
Currently the whole table is getting re-rendered
This degrades the performance dramatically specially when you have 100's or 1000's of records.

**Current options** - [Virtual List](https://ant.design/components/table/#components-table-demo-virtual-list)
Yes you can use virtual list looks like now, but that would probably mean rewriting the table because it is not a small change, the example in docs shows this as a much bigger change and a completely different way to write your tables. It's not just a prop setting.

**Real use cases**
In some cases we want to show 1000 records per page or more, so that people are able to select more records and perform that action in bulk instead of in smaller sets. Not sure how virtual list affects that.
Also we may have extended the table and added custom features.
This rendering re-renders all attached components, when really that may not be needed.

**Bottom line**
It does not take away from the fact that **NOT** all rows should be re-rendered when only one row changes

| Environment | Info |
|---|---|
| antd | 4.2.0 |
| React | 16.13 |
| System | Mac OSX |
| Browser | Safari 13.1, Chrome 80.0 |

**Comment by DeadManPoe** [comment]
My company needs this too

**Comment by smddzcy** [comment]
Having the same issue (this and #10431), even a table with 100 rows is very slow when the row selection feature is enabled.

**Comment by zombieJ** [comment]
Table can not tell what will use in column `render` function, it always need fully render by default:
```tsx
const [timestamp, setTimestamp] = React.useState(Date.now());

const columns = [{
  render: () => timestamp,
}];

return <>
    <Table columns={columns}  />
    <button onClick={() => setTimestamp(Date.now())} />
  </>;
```

But you can use `components` prop to cache the render: https://codesandbox.io/s/selection-and-operation-ant-design-demo-buogh?file=/index.js

**Comment by sowka1995** [comment]
> Table can not tell what will use in column `render` function, it always need fully render by default:
> 
> ```tsx
> const [timestamp, setTimestamp] = React.useState(Date.now());
> 
> const columns = [{
>   render: () => timestamp,
> }];
> 
> return <>
>     <Table columns={columns}  />
>     <button onClick={() => setTimestamp(Date.now())} />
>   </>;
> ```
> 
> But you can use `components` prop to cache the render: https://codesandbox.io/s/selection-and-operation-ant-design-demo-buogh?file=/index.js

There is some kind of bug in this cache example.
Select row number 3 then select row number 2, a then unselect row number 3 and both second and third row are being unselected.

**Comment by StallionV** [comment]
I will be **Blunt**.
@zombieJ the example you provided is buggy, I think we are more in a hurry to close issues than to actually address the core problems. In the example you provided when you toggle rows one after another you see a vey buggy behavior since it seems to select unselect other rows also.

We also are undermining the fact that the table has to be designed from a perspective of performance, simply stating that 

> Table can not tell what will use in column render function, it always need fully render by default:

points to the fact that we are not open to critical suggestions or feedback. Not rendering a whole table on row change is something that should have been part of the design from start. It will be great if Ant team acknowledges that and tries to rectify it rather than brush it under the carpet and keep closing these issues saying this is how it was designed.

**Comment by DeadManPoe** [comment]
@StallionV I totally agree with you

**Comment by jkdlouis** [comment]
@zombieJ This issue should not be closed until a permanent solution is provided, can you at please provide a work around since the example you provided is broken?

**Comment by StallionV** [comment]
> @zombieJ This issue should not be closed until a permanent solution is provided, can you at please provide a work around since the example you provided is broken?

100% agree with @jkdlouis

**Comment by zombieJ** [comment]
Thanks for report. Let me reopen this.

**Comment by zombieJ** [comment]
Update: Check for the internal logic. Selection deps on the keys which was cached by `useMemo` this should be update to support memo component.

**Comment by zombieJ** [comment]
Columns will support `shouldCellUpdate` function in next minor version.

**Comment by StallionV** [comment]
@zombieJ @afc163 can one of you please add a good example as well for using `shouldCellUpdate`

**Comment by nikhilsarvaiye** [comment]
Hi, can you anyone please let me know if this available for use, if yes npm version we can use, as this has become important deciding factor, thanx

**Comment by yoyo837** [comment]
> Hi, can you anyone please let me know if this available for use, if yes npm version we can use, as this has become important deciding factor, thanx

v4.3.0 in the future.

**Comment by StallionV** [comment]
> > Hi, can you anyone please let me know if this available for use, if yes npm version we can use, as this has become important deciding factor, thanx
> 
> v4.3.0 in the future.

But when in future ?

**Comment by yoyo837** [comment]
There are no regular plans for minor and major releases.

**Comment by StallionV** [comment]
Released :)

**Comment by almeidapaulooliveira** [comment]
The documentation is very poor is this feature. It will be great to see some example of using. Thank you.

**Comment by StallionV** [comment]
@yoyo837 @zombieJ @afc163 please can you update documentation with examples

**Comment by yeomann** [comment]
Doesn't even make sense, i have columns in this way
![image](https://user-images.githubusercontent.com/5970751/88768903-66deff00-d184-11ea-9739-7be0154cb457.png)

try to implement above shouldCellUpdate feature, again it trigger everywhere. Feels like this method is not doing anything, i think it made those check boxes more slower.

Here are the console results of thousands records
![image](https://user-images.githubusercontent.com/5970751/88769100-aa396d80-d184-11ea-84b7-4078d93a857e.png)

what a pity, feels like wasted time now

**Comment by StallionV** [comment]
This is a repeated request from everyone @yoyo837 @zombieJ @afc163 please can you update documentation with examples for shouldCellUpdate

**Comment by zombieJ** [comment]
Patch the test case: https://github.com/react-component/table/blob/659158f0aa8b844c39a66c808c425d72865e327d/tests/Cell.spec.tsx#L18

Pls feel free to provide reproduce if it exist some uncatched case.

**Comment by yeomann** [comment]
![image](https://user-images.githubusercontent.com/5970751/88904537-decd2800-d25d-11ea-9157-73311d10a7e1.png)

so same logic and i am at `    "antd": "^4.5.1",`
on first render i see this 
![image](https://user-images.githubusercontent.com/5970751/88904642-ff957d80-d25d-11ea-8b45-1bf0a48577f8.png)
but when i scroll down and come up the ID is still increasing
![image](https://user-images.githubusercontent.com/5970751/88904681-0d4b0300-d25e-11ea-971e-b679cfbdf806.png)

This indicates to me that re-rendering is still happening.

just to add more notes, I am using ` components={vt}`
```
 const [vt] = useVT(() => {
    return {
      id: 0,
      // overscanRowCount: 10,
      scroll: { y: 500, x: '100vw' },
      debug: false,
    }
  }, [])
```
and useVT from `import { useVT } from 'virtualizedtableforantd4'`


The performance for the selecting checkbox is increased by using `virtualizedtableforantd4` but why we are seeing this ID increase on scroll. I have no idea.

Can someone highlight what more needs to be done to stop this re-rendering issue.

**Comment by zombieJ** [comment]
@yeomann  could you provide a codesandbox about your reproduce? thx

**Comment by yeomann** [comment]
check this out https://codesandbox.io/embed/selection-and-operation-ant-design-demo-4co4k?fontsize=14&hidenavigation=1&theme=dark

i couldn't add virtualizedtableforantd4 as codesandbox was giving error, none the less go through the pagination and you will see that ID is increasing all the times which indicates the re-render @zombieJ

**Comment by zombieJ** [comment]
> check this out https://codesandbox.io/embed/selection-and-operation-ant-design-demo-4co4k?fontsize=14&hidenavigation=1&theme=dark
> 
> i couldn't add virtualizedtableforantd4 as codesandbox was giving error, none the less go through the pagination and you will see that ID is increasing all the times which indicates the re-render @zombieJ


```diff
--- "antd": "4.1.5",
+++ "antd": "^4.5.0",
```

**Comment by yeomann** [comment]
did you even check the codesandbox? its same even with 4.5.2 https://codesandbox.io/s/selection-and-operation-ant-design-demo-4co4k?from-embed

**Comment by StallionV** [comment]
Please reopen this issue until it is completely resolved

**Comment by zombieJ** [comment]
> did you even check the codesandbox? its same even with 4.5.2 https://codesandbox.io/s/selection-and-operation-ant-design-demo-4co4k?from-embed

` ID is increasing all the times` is not longer exist in your demo when update with `4.5.2`. Pls check again.

**Comment by yeomann** [comment]
Screenshot 1: 
![image](https://user-images.githubusercontent.com/5970751/89503038-a0d88280-d7ce-11ea-852b-6e84d843466f.png)

Click on any page from pagination.
Come back to first page.
![image](https://user-images.githubusercontent.com/5970751/89503085-b5b51600-d7ce-11ea-8167-473c6f69eea6.png)

As you can see ID is increased.
This indicate the Re-render.

Also look at the Console. you can see Re-render

![image](https://user-images.githubusercontent.com/5970751/89503171-d8dfc580-d7ce-11ea-9d37-c9323cf3c1ab.png)


Please check again same codesandbox
https://codesandbox.io/s/selection-and-operation-ant-design-demo-4co4k?from-embed

**Comment by zombieJ** [comment]
`shouldCellUpdate` compares the cell render, not row render. When you click different page, record of each row is different which trigger the `column.render` function. You can change it to `shouldCellUpdate() => false` and try again. `shouldCellUpdate` works as expect in your case.

**Comment by yeomann** [comment]
how can we stop re-render of ROWS?? @zombieJ 
we don't want the re-render the rows / cells whatever.

is it possible to stop re-rendering the rows, i don't understand what is the logic of re-rending rows? maybe you have this logic for some layout re-rendering. but how can we stop please?

**Comment by zombieJ** [comment]
Could you tell which case you want to stop re-render row?

**Comment by DeadManPoe** [comment]
@zombieJ An important use case can be the following: I don’t really want to re-render all the rows of the current page of a table when I select one row.

**Comment by lorenzocestaro** [comment]
I've been following this issue for a good while. I'm working on a product based mainly on a table and I did not like the fact that all rows rendered each time selected a row. I must say though that the issue is indeed solved. In my specific case, since all the cells in my table display read-only static data, I just added to all column properties `shouldCellUpdate: () => false`. In this way, when you interact with a row (i.e. select it in my case) the table does not rerender.

@yeomann, you can't expect no renders when you change pagination. You are changing your view to see the next batch of rows, of course they need to be rendered. And if you go back to the previous page the rows still need to be rendered, they were unmounted from the DOM and now need to be drawn again. Even using the virtualized table you cannot expect to not have renders, even more so in this case. The virtualized table improves performance by dynamically rendering only the rows that are visible at any single moment on your page. This means that when a row is scrolled out of view, its node gets removed from the DOM, and if you scroll back up that row needs to be rendered again.

**Comment by thekennysong** [comment]
Is there any update on this issue?

**Comment by amodes** [comment]
Thanks  @[lorenzocestaro](https://github.com/lorenzocestaro)

As I have dynamic data I am adding to every column in the table
```js
shouldCellUpdate: (record, prevRecord) => !_.isEqual(record, prevRecord);
```

**Comment by wenlittleoil** [comment]
> Thanks @[lorenzocestaro](https://github.com/lorenzocestaro)
> 
> As I have dynamic data I am adding to every column in the table
> 
> ```js
> shouldCellUpdate: (record, prevRecord) => !_.isEqual(record, prevRecord);
> ```



> Thanks @[lorenzocestaro](https://github.com/lorenzocestaro)
> 
> As I have dynamic data I am adding to every column in the table
> 
> ```js
> shouldCellUpdate: (record, prevRecord) => !_.isEqual(record, prevRecord);
> ```

Thanks a lot, it works for me.

**Comment by tobiasBora** [comment]
For people using redux, I used successfully:
```
shouldCellUpdate: (record, prevRecord) => !shallowEqual(record, prevRecord)
```
